Make table cards clickable

seancolsen opened the pull request:
Fixes #1978

## Before this PR

- Navigating to a table page required the user to click on the "Go to Table" link within the table card.

    ![image](https://user-images.githubusercontent.com/42411/211919630-75e6eb77-87cd-4bc8-9dc3-45e2f3bfc965.png)

    We want to be able to click on other areas of the card to access the hyperlink, like the table name.

- We had a PR #2003 from @LFLATH to fix this issue, but the PR had some problems, which stemmed from design shortcomings.

- @ghislaineguerin and I had a call where we discussed some design changes to the table card. She documented the [new design in Figma](https://www.figma.com/file/xHb5oIqye3fnXtb2heRH34/Styling?node-id=4425%3A28850&t=piHbh9OqhGLLjRJ6-0).

- Using a horizontal ellipses for the menu trigger was a bit confusing when juxtaposed with the ellipses used for a truncated table name

    ![image](https://user-images.githubusercontent.com/42411/211919798-ca0336b0-287a-4bd0-bfc3-e70ef0fce279.png)

## Changes in this PR

- I created a new `HyperlinkCard` component.

    I'd like to eventually use it for the schema cards that we display on the Database Page too, but I didn't want to get too carried away with this PR.

    For consistency, I made the styling of the `HyperlinkCard` to mirror the styling used for the Schema Card.

- I modified the menu icon to use a vertical ellipses instead of a horizontal ellipses, and I made this change for the schema cards too.

- The entire table card can now be hovered

    ![image](https://user-images.githubusercontent.com/42411/211916624-002a2c1c-75c6-4688-963e-d92496f0f68f.png)

- Then menu trigger extends to the full edges of the card, whereas previously the card had some padding which would have created a weird hyperlink click area between the menu trigger and the edge of the card.

    ![image](https://user-images.githubusercontent.com/42411/211916771-465aa6c2-649b-4742-958e-474ee275fb95.png)

- I moved "Find a Record" into the menu.

    ![image](https://user-images.githubusercontent.com/42411/211917055-35c85cf7-075b-4980-a446-5af2d1868f29.png)

    @kgodey @rajatvijay and I had some discussion [in Matrix](https://matrix.to/#/!xInuTkBwjXZXYatIlm:matrix.mathesar.org/$1UtRqvDIJbgY_vgOQux8LGkZ3fKc1lWfyAY_-nh6fpg?via=matrix.mathesar.org&via=matrix.org) regarding discoverability of the Record Selector feature. The change in this PR _does_ make the Record Selector slightly less discoverable here, but I think that's okay.

- Truncated values are still visible on hover.

    ![image](https://user-images.githubusercontent.com/42411/211916891-05f53d38-18cb-4389-8f5f-69c12d95cd2c.png)

    This was one of the problems with #2003 from @LFLATH. I had to take a bit of a fundamentally different approach here to fix that issue.

- I removed the `max-width` on the table cards. That was resulting in this minor anesthetic issue with spacing:

    ![image](https://user-images.githubusercontent.com/42411/211922833-81d864ea-3cd8-4105-9f9c-f0f3f26b79fd.png)

    I think this spacing is better:

    ![image](https://user-images.githubusercontent.com/42411/211923039-ddd0d172-6ea2-455b-b5a8-28251e9be323.png)


## Issues

- @ghislaineguerin made this [Figma design](https://www.figma.com/file/xHb5oIqye3fnXtb2heRH34/Styling?node-id=4425%3A28850&t=piHbh9OqhGLLjRJ6-0) which uses separate background colors for the card vs the page. 

    ![image](https://user-images.githubusercontent.com/42411/211917975-68522a27-60a0-438f-bc3a-c4c9fc3d6adc.png)

    I really like this design, but fully implementing it will require more far-reaching changes. For example, our `TabContainer` component requires that the tabs and the tab content both be placed in the same element, making it tricky to give the tabs the tan background shown above. I'd like to handle this later and leave it to @ghislaineguerin to prioritize/track.

- I don't love the styling when the dropdown trigger button is focused

    ![image](https://user-images.githubusercontent.com/42411/211916079-fb239c24-be49-440a-91c4-763cdee76e76.png)

    This state results when you click on the menu trigger. With the menu open the focus indicator isn't super apparent. But if you click on the trigger again (closing the menu, while still keeping the trigger focused) then you enter the state in the screenshot above.

    This is tricky to change because of the usage of `!important` within our component library. We already have #1042 to remove usage of `!important`, so I'd rather leave this as-is for now. This is minor enough that I don't think it's worth tracking via a separate issue. I just wanted to mention it in case someone noticed it during review.


## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

seancolsen commented:
@kgodey this PR contains some design changes that you might like to look over.

rajatvijay commented on mathesar_ui/src/components/HyperlinkCard.svelte:
```diff
@@ -0,0 +1,106 @@
+  let isHoveringMenuTrigger = false;
+</script>
+
+<div class="hyperlink-card" class:hovering-menu-trigger={isHoveringMenuTrigger}>
```
Since this is a common component, when used in other places might end up with invalid markup (button or links inside the wrapper link). This could happen majorly because its uses `slots` to render `top` and `bottom` contents and the consumer can pass a button or a link in those slots. 

Wouldn't it be better if we used something like this: https://svelte.dev/repl/40411730765d45c3a1b46c665562be70?version=3.55.1

Few minor things on the current implementation: 

1. I see a focus styling missing on the whole card but maybe this is something that we can handle when doing keyboard interactions. 
2. The hover styling of the vertical three dots menu doesn't seem to respect the border-radius of the card. This is something that you might want to handle in this PR itself.
<img width="361" alt="Screenshot 2023-01-12 at 10 04 40 AM" src="https://user-images.githubusercontent.com/11032856/211976993-ced7e508-0d69-462a-b3d3-78ab8aa0425d.png">

seancolsen replied:
> when used in other places might end up with invalid markup

- There are a number of things about your REPL code which confuse me. I think it might be best for us to hop on a call to hash this out. This table card has some nuances and tricky requirements. For example

    - (A) The whole card area needs to be clickable as a hyperlink, except the menu trigger.
    - (B) When hovering a truncated description, the user should see the popover from the truncation component.

    The combination of (A) and (B) lead me to the conclusion that the description needs to be a child of the hyperlink. I'm pretty sure that in your REPL we wouldn't be able to hover the table description to see its full value

    If this critique is something we could handle in a separate PR it might be nice, just so we can get this fix in before the release.

> focus styling missing on the whole card

- Yeah, I'd rather fix that separately from this PR. We have lots of other areas of the app that need similar improvements. I've opened #2248 to track that work.

> The hover styling of the vertical three dots menu doesn't seem to respect the border-radius of the card. 

- I've pushed ef85a568bbf40ea0f26c644201ce9d4713f5cb7b to fix that. It's weird to me that I wasn't seeing that hover problem when originally working on the PR (as is evident in my screenshots above). But I was indeed able to reproduce it now. Oh well! This fix ought to do the trick I think.

kgodey commented:
>  The change in this PR does make the Record Selector slightly less discoverable here, but I think that's okay.

I think more important that discoverability, this makes finding a record a 2-click operation instead of a 1-click operation. In our library use case, when a patron comes in, the librarian finds them from the schema page. This is a frequent operation, and the extra clicks add up over time. Also, the click target is much smaller, which shouldn't be the case for common oper

Additionally, here's some design issues I found:

1. Hovering over a long description obscures the name of the table. I think it would make more sense for the tooltip to be on the bottom rather than the top.
<img width="368" alt="Screen Shot 2023-01-12 at 4 35 57 PM" src="https://user-images.githubusercontent.com/287034/212186713-9f2c003c-5f48-4130-a2e8-75a9041cb19e.png">

2. We can end up with different size cards for tables without a description. I think this looks weird. I don't have a particular solution to suggest, but I did like that the previous design had a uniform size.
<img width="664" alt="Screen Shot 2023-01-12 at 4 37 11 PM" src="https://user-images.githubusercontent.com/287034/212186887-4f9dd950-1c62-43f5-9068-14bdc9485f9b.png">

---

After playing around it with it locally, I prefer our current (in `master`) table cards over the ones in this PR. I do like that the entire card is clickable, but otherwise it seems more inconsistent / harder to use.

kgodey commented:
Another design issue – the old design clearly shows when a table has been imported and the data needs to be confirmed. I haven't looked to see what we're doing in this PR, but it should be as clear.

<img width="331" alt="Screen Shot 2023-01-12 at 5 02 49 PM" src="https://user-images.githubusercontent.com/287034/212190881-324c3c56-e9e2-4533-ac7f-ace38b2ee532.png">

seancolsen commented:
@kgodey how would you like to proceed here? Being that you authored #1978 requesting that the table cards be clickable, would you prefer that we abandon that ticket and close this PR? Should we have another round of design? Ghislaine and I already spent some time together on a call to formulate the design that I implemented in this PR. Should we have a design session with the three of us together? I don't have an opinion. I'm fine with the design in `master` and fine with the design in this PR.

As I see it issues `1` and `2` are also present in `master`. E.g....

![image](https://user-images.githubusercontent.com/42411/212190653-39059c37-66cc-419b-a98f-b20fbc2ea196.png)

I can fix `1`. But `2` is tricky. I played with making the cards vertically shorter but that looked even worse. 

I agree that discoverability of the record selector could be a problem. I'm open to other suggestions here.

seancolsen commented:
Oh good point about the unconfirmed imports. This PR _does need additional work_ to handle that! I forgot about that.

kgodey commented:
@seancolsen Ideally, I'd like to update this PR to:
- Fix issue 1 with the tooltip positioning
- bring the "Find a Record" button back, perhaps as a single full width button similar to "Confirm Imported Data"
- Bring the "Confirm Imported Data" button back as well
- For issue 2, `master` doesn't have the of some cards being tall and some cards being short if they are description-less in different rows (see "Author" and "Paper-Author Map" in my screenshot). I think we should EITHER:
  - Make the cards the same height regardless of description being present and live with the empty extra space (current `master`'s solution). 
  - Have a masonry style layout for the cards (I'm guessing this will take a lot more work and we should do the first one).

pavish commented:
@seancolsen Looks like there are UX changes required here, so I'm going to hold off code review until they are compelete.

seancolsen commented:
@kgodey Ok how about this.

https://user-images.githubusercontent.com/42411/212226068-33292007-e1c4-456d-8395-9bfc535eb840.mp4

- In 74754ee65904b6625c31fb85ec466253fcc36b7c I fixed the issue with the tooltip positioning.

- In defb122f847a7e21a0c3f0859f92807d34d50d5f I combined TableCard and HyperlinkCard, which I hope will address @rajatvijay's concerns (because TableCard can not compose other component as HyperlinkCard could have). This card is getting sufficiently complex now that the HyperlinkCard abstraction doesn't feel worth it anymore anyway.

- In db38b14eabfd6e5d1ebbdb95514f76b31fa6ca04 I adjusted some styles, including some improvised changes...

    - I added a "Find a Record" button to the bottom of the card (and removed that entry from the dropdown menu)

    - I gave unconfirmed imports a distinct style. Notably, I didn't want the "Confirm Imported Data" to look quote the same as "Find a Record" because "Confirm Imported Data" is _the same link as the card_ whereas "Find a Record" is a button separate from the card's hyperlink.

    - I decreased the text size of "Find a Record". This is to address a concern of mine about the previous design where, when scanning the list of tables I would sometimes get lost within other text. With this change, the table name stands out more.

    - I also decreased the text size of the description for the same reason as above. This change also has the benefit of reducing the vertical size discrepancy between a card with no description and a card with a full description.

- I did _not_ make any changes to address this concern `2` of Kriti's.

    > master doesn't have the of some cards being tall and some cards being short if they are description-less in different rows

    @kgodey, as far as I can tell, the behavior you're critiquing is also present in `master`. Here's what I observe in `master`

    ![image](https://user-images.githubusercontent.com/42411/212228041-1814725b-b6d3-48b8-89e9-6c709f4347bf.png)

    Notice that the "Authors" card is short, the "Items" card is somewhat taller, and the "Publications" card is taller still. The difference in height is less apparent due to the additional constant factor of the bottom section of the card. I'm hoping that my changes above will bring this PR into close-enough parity with `master` to alleviate your concern.

    I agree that a masonry layout would be ideal here. Masonry layout is [coming](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Grid_Layout/Masonry_Layout) to browsers natively, but it'll be a while before it's viable. So you're correct that it wouldn't be quick to implement.

    I'd rather not make all cards as tall as the tallest possible card for the following reasons:

    - This is tricky to implement cleanly.
    - I predict that many people will have schemas with lots of tables that don't have descriptions. I think a whole grid of cards that are taller than they need to be will look awkward and decrease usability by requiring more scrolling to find a table.

    If we do still want to seek changes to address concern `2`, I'm hoping we can continue discussion/development in a separate PR.

kgodey commented:
I just watched the video, I did not play with it locally, but I think these changes satisfy all my concerns, nice work @seancolsen.  I'm unassigning myself from this PR. 

I'd suggest changing the "Confirm Imported Data" text to "Needs Confirmation" or something like that since it's no longer an action button.

> Notice that the "Authors" card is short, the "Items" card is somewhat taller, and the "Publications" card is taller still.

I didn't realize that. Maybe it's just that the original height difference in this PR was so different that it was jarring.

rajatvijay commented:
@seancolsen looks good to me too. Since the `HyperlinkCard` does not exist anymore I will put my suggestion as an item for the next FE meeting as a low-priority item as we do not have a real use case for this now.

seancolsen commented:
@kgodey I pushed 6bc0ed58ca31066580bc929507ccf074c00f8df7 changing the text to "Needs Import Confirmation". That seemed a bit clearer than "Needs Confirmation".

![image](https://user-images.githubusercontent.com/42411/212370847-b9392542-495e-467d-b3fa-d9402ba1df83.png)

kgodey commented:
Looks good, thanks @seancolsen

pavish reviewed (approved):
Looks good
